// @datocms/cli/lib/commands/profile/remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const cli_utils_1 = require("@datocms/cli-utils");
class Command extends cli_utils_1.DatoConfigCommand {
    async run() {
        var _a;
        const { PROFILE_ID: profileId } = this.parsedArgs;
        if (!this.datoConfig) {
            this.log(`Config file not present in "${this.datoConfigRelativePath}", skipping operation`);
            return;
        }
        if (!(profileId in this.datoConfig.profiles)) {
            this.log(`Config file does not contain profile "${profileId}", skipping operation`);
            return;
        }
        await this.saveDatoConfig({
            ...this.datoConfig,
            profiles: Object.fromEntries(Object.entries(((_a = this.datoConfig) === null || _a === void 0 ? void 0 : _a.profiles) || {}).filter(([key]) => key !== profileId)),
        });
    }
}
exports.default = Command;
Command.description = 'Remove a profile from DatoCMS config file';
Command.args = [
    {
        name: 'PROFILE_ID',
        description: 'The name of the profile',
        required: true,
    },
];
