// @datocms/cma-client/dist/esm/generated/resources/SearchResult.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import * as Utils from '@datocms/rest-client-utils';
import BaseResource from '../../BaseResource';
var SearchResult = /** @class */ (function (_super) {
    __extends(SearchResult, _super);
    function SearchResult() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    /**
     * Search for results
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/search-result/instances
     */
    SearchResult.prototype.list = function (queryParams) {
        return this.rawList(queryParams).then(function (body) {
            return Utils.deserializeResponseBody(body);
        });
    };
    /**
     * Search for results
     *
     * Read more: https://www.datocms.com/docs/content-management-api/resources/search-result/instances
     */
    SearchResult.prototype.rawList = function (queryParams) {
        return this.client.request({
            method: 'GET',
            url: "/search-results",
            queryParams: queryParams,
        });
    };
    SearchResult.TYPE = 'search_result';
    return SearchResult;
}(BaseResource));
export default SearchResult;
